sdk/rust/src/http.rs

```diff
@@ -15,13 +17,165 @@ pub use super::wit::wasi::http::types::*;
 /// is no need for streaming bodies.
 pub struct Request {
     /// The method of the request
-    pub method: Method,
-    /// The path together with the query string
-    pub path_and_query: String,
+    method: Method,
+    /// The uri for the request
+    ///
+    /// The first item is set to `None` if the supplied uri is malformed
+    uri: (Option<hyperium::Uri>, String),
     /// The request headers
-    pub headers: Vec<(String, Vec<u8>)>,
+    headers: HashMap<String, HeaderValue>,
```

Comment:
I know this was my suggestion but since we aren't exposing the `Map`ness of this directly we could retain a bit of HTTP compliance by storing this as a `Vec` and e.g. `self.headers.iter().rev().find()`ing in `self.header`.

Reply:
Ah good catch - do you mean because a header can have more than one value? How about `HashMap<String, Vec<HeaderValue>>`? It seems worth it to save having to iterate through all headers to find a specific one.

Reply:
Yeah, multiple values are _almost_ uncommon, but basically mandatory for `set-cookie`.

Reply:
uuuuugghhhh we're probably better off using `http::HeaderMap` internally :neutral_face:

Reply:
`HeaderMap` is annoying in that it checks the bytes in `HeaderValue` for validity which may be more spec compliant but means we have to deal with that instead of letting it be someone else's problem.

Reply:
Yeah OK. Well, we do still need to handle case-sensitivity. Downcasing is probably the simplest way to manage that.